### PR TITLE
tests: add builder block assembly and overflow tests

### LINKED ISSUE #19629: Remove ExecV3: Create faked TxnProvider for deterministic builder testing
### Phase 1a: Faked TxnProvider

**New file**: `execution/builder/builderstages/exec_test.go`

Create a `fakeTxnProvider` implementing `txnprovider.TxnProvider` that returns pre-signed transactions deterministically, without needing a full txpool. This is the foundation for unit-level builder tests.

```go
type fakeTxnProvider struct {
    txns []types.Transaction
    idx  int
}
func (f *fakeTxnProvider) ProvideTxns(ctx context.Context, opts ...txnprovider.ProvideOption) ([]types.Transaction, error) {
    o := txnprovider.ApplyProvideOptions(opts...)
    remaining := f.txns[f.idx:]
    count := min(o.Amount, len(remaining))
    result := remaining[:count]
    f.idx += count
    return result, nil
}
```

This gives deterministic transaction ordering and is reusable for any builder test that needs to control exactly which transactions the builder sees.

### LINKED ISSUE #19630: Remove ExecV3: Extend ExecModuleTester block assembly tests
### Phase 1b: Extend ExecModuleTester tests

**File**: `execution/execmodule/exec_module_test.go`

Extend existing `TestAssembleBlock` with post-assembly state assertions and add new test cases:

- **Balance verification**: Query sender/receiver balances after FCU, assert correctness
- **Nonce verification**: Assert sender nonce incremented by number of txns
- **`TestAssembleEmptyBlock`**: Build block with no pending txns — only system calls modify state. Simplest case, catches txNum off-by-one errors.
- **`TestAssembleBlockWithContractCreation`**: Deploy contract via txpool, verify code exists after block assembly

Key existing helpers to reuse:
- `assembleBlock()` / `getAssembledBlock()` (exec_module_test.go:362-380)
- `insertValidateAndUfc1By1()` (exec_module_test.go:334-360)
- `addTwoTxnsToPool()` (exec_module_test.go:178-202)
- `ExecModuleTester` with `WithTxPool()`, `WithChainConfig()` options

These tests validate correctness through the existing code path and will then validate the refactored code produces identical results.

### LINKED ISSUE #19631: Remove ExecV3: Add EngineApiTester builder integration tests
### Phase 1c: EngineApiTester integration tests

**New file**: `execution/engineapi/engine_api_builder_test.go`

Full-stack Engine API → builder → validation tests:

- **`TestEngineApiBuiltBlockStateMatchesValidation`**: Submit transfer via `Transactor`, build via `MockCl.BuildCanonicalBlock()`, query balances/nonces via `RpcApiClient`. Key assertion: `ValidateChain` passes (builder's state root matches ExecV3 validation).

- **`TestEngineApiMultiBlockSequence`**: Build 5 consecutive blocks with txns, verify each is accepted. Tests SharedDomains state across multiple builder invocations.

- **`TestEngineApiEmptyBlockProduction`**: Build block with no pending txns. Verifies system calls produce correct state root.

- **`TestEngineApiBuiltBlockReorgRecovery`**: Build block A, then build block B on a different fork. Verify builder handles SharedDomains rollback.

Key existing infrastructure:
- `engineapitester.DefaultEngineApiTester(t)` (engine_api_tester.go)
- `MockCl.BuildCanonicalBlock()` (mock_cl.go)
- `Transactor.SubmitSimpleTransfer()` (transactor.go)
- `TxnInclusionVerifier.VerifyTxnsInclusion()` (txn_inclusion_verifier.go)

### LINKED ISSUE #19645: Phase 1f: Block overfill and filtering rule tests
## Context
Part of #19318 — builder test infrastructure.

## What
Add tests that exercise block capacity overflow (transactions spilling across blocks) and transaction filtering rules (nonce validation, multiple senders, mixed tx types, gas accounting).

### EngineApiTester tests
- `TestEngineApiBlockGasOverflowSpillsToNextBlock` — custom 150K gas limit, 10 transfers split across 2 blocks
- `TestEngineApiSequentialNonceAdvancement` — nonce validation across sequential blocks
- `TestEngineApiMultipleSendersInBlock` — interleaved sender filtering with 2 funded accounts
- `TestEngineApiHighGasContractsFillBlock` — custom 200K gas limit, contract deploys + transfers cause spillover

### ExecModuleTester tests
- `TestAssembleBlockGasOverflow` — custom 150K gas limit, 10 txns via txpool split across 2 assembled blocks
- `TestAssembleBlockMixedTxTypes` — transfers + contract creation in same block

### Filtering rules exercised
- Gas pool exhaustion (block gas limit overflow)
- Nonce advancement and validation
- Multiple sender accounts in same block
- Mixed tx types (transfers + contract creation)
- High gas contract deploys filling blocks

Parent: #19318

### PR DESCRIPTION
**Pull request by mh0lt**
## Summary
- Add `fakeTxnProvider` test double implementing `txnprovider.TxnProvider` for block builder unit tests
- Add ExecModuleTester tests: empty block assembly, multi-block state verification, contract creation, gas overflow, mixed tx types
- Add EngineApiTester integration tests: state validation, multi-block sequence, empty block, contract deploy+call, reorg recovery, gas overflow spillover, nonce advancement, multiple senders, high-gas contract block filling

## Filtering rules exercised
- Block gas limit overflow with transaction spillover across blocks
- Sequential nonce advancement and validation
- Multiple sender accounts interleaved in same block
- Mixed transaction types (transfers + contract creation)
- High-gas contract deploys filling blocks with spillover

## Issues
Fixes #19629 — Phase 1a: Faked TxnProvider
Fixes #19630 — Phase 1b: Extend ExecModuleTester tests
Fixes #19631 — Phase 1c: EngineApiTester integration tests
Fixes #19645 — Phase 1f: Block overfill and filtering rule tests

Parent issue: #19318

## Test plan
- [x] `go test ./execution/builder/builderstages/... -run "TestFakeTxnProvider" -v` (2 tests pass)
- [x] `go test ./execution/execmodule/... -run "TestAssemble" -v` (6 new + 2 existing tests pass)
- [x] `go test ./execution/engineapi/... -run "TestEngineApi" -v` (9 new + 2 existing tests pass)
- [x] `make erigon` — builds clean
- [x] `make lint` — 0 issues

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: execution/execmodule/exec_module_test.go [hunk 1]
```diff
@@ -399,3 +401,340 @@ func retryBusy[T any](ctx context.Context, f func() (T, bool, error)) (T, error)
...
+
+	block, err := getAssembledBlock(ctx, exec, payloadId)
+	require.NoError(t, err)
+	require.Equal(t, uint64(2), block.NumberU64())
+	require.Len(t, block.Transactions(), 1)
+
+	// Insert + validate + FCU — validates state root which proves
+	// contract deployment was executed correctly.
+	err = insertValidateAndUfc1By1(ctx, exec, []*types.Block{block})
+	require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (mh0lt) [reply execution/execmodule/exec_module_test.go:565]: This is a nice to have fix that won't actually compile.  The check that the stateroot is the same as the built root is intended as the pass/fail criteria
